### PR TITLE
Introduce extensions to OracleLinux/7 project

### PR DESCRIPTION
**Pull request by AmedeeBulle**
Make the OracleLinux/7 project extensible to avoid proliferation of _simple_ sample projects.

--
Signed-off-by: Philippe Vanhaesendonck <philippe.vanhaesendonck@oracle.com>

### CODE REVIEW

File: OracleLinux/7/.env [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Oracle Linux 7 configuration file
+#
+# Requires vagrant-env plugin
+#
+# This file will be overwritten on updates, it is recommended to make changes
+# in .env.local
+
+# Extensions
+# To extend the box, uncomment one of the following:
```
Comment (Djelibeybi) [inline OracleLinux/7/.env:9]: I think this might be a bit confusing as users would probably uncomment the "`LAMP server`" heading/break.

File: OracleLinux/7/README.md [hunk 2]
```diff
@@ -1,36 +1,89 @@
 # ol7-vagrant
-A vagrant box that provisions Oracle Linux automatically, using Vagrant, an Oracle Linux 7 box and a shell script.
+
+A vagrant project that provisions Oracle Linux automatically, using Vagrant, an Oracle Linux 7 box and a shell script.
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:3]: This could perhaps be better phrased as "A Vagrant project to automatically build an Oracle Linux 7 virtual machine using either VirtualBox or libvirtd with optional extras including a LAMP stack or container runtime."

File: OracleLinux/7/README.md [hunk 3]
```diff
@@ -1,36 +1,89 @@
...
 vagrant plugin install <name>...
 ```
 
+## Extending the project
+
+This project can easily be extended by running additional scripts during provisioning and optionally expose guest ports on the host.
+
+Environment variables are used to pass the parameters to the Vagrantfile:
+
+- `EXTEND`: coma separated list of extensions.
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:46]: Typo: should be "comma"

File: OracleLinux/7/.env [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+# LAMP server
+# EXTEND=lamp
+# EXPOSE=8080:80
+
+# Docker engine
+# EXTEND=docker-engine
```
Comment (Djelibeybi) [inline OracleLinux/7/.env:16]: Should this not be `EXTEND=$EXTEND,docker-engine` to allow for multiple extensions? I would also rename it to `container-runtime` (see below for the more detailed reason).

File: OracleLinux/7/README.md [hunk 5]
```diff
@@ -1,36 +1,89 @@
...
 ```
 
+## Extending the project
+
+This project can easily be extended by running additional scripts during provisioning and optionally expose guest ports on the host.
+
+Environment variables are used to pass the parameters to the Vagrantfile:
+
+- `EXTEND`: coma separated list of extensions.  
+   For each extension a script `<extension>.sh` is run. Scripts must be located in the `scripts` or `scripts.local` directory.
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:47]: "For each specified extension a script named `<extension>.sh` is automatically run during provisioning."

File: OracleLinux/7/README.md [hunk 6]
```diff
@@ -1,36 +1,89 @@
...
+- `EXPOSE`: coma separated list of ports to expose in the format: `<host port>:<guest port>`
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:48]: Typo: "comma"

File: OracleLinux/7/README.md [hunk 7]
```diff
@@ -1,36 +1,89 @@
...
+
+```shell
+EXTEND=lamp EXPOSE=8080:80 vagrant up
+````
+
+Alternatively, if the `vagrant-env` plugin is installed variables can be defined in the `.env` or `.env.local` files.
+
+## Sample extensions
+
+### LAMP
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:60]: It's probably best to define what a LAMP stack is at least once.

File: OracleLinux/7/README.md [hunk 8]
```diff
@@ -1,36 +1,89 @@
...
+
+Provisions an Oracle Linux LAMP solution.
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:62]: I'd replace "solution" with "stack"

File: OracleLinux/7/README.md [hunk 9]
```diff
@@ -1,36 +1,89 @@
...
+EXTEND=lamp
+EXPOSE=8080:80
+```
+
+Guest port "80" (Apache) is redirected to Host port "8080".  
+Once ready, you can test it by opening following URL on your Host OS: http://localhost:8080/info.php
+
+### Docker engine
+
+Installs and configures Docker engine with Btrfs as storage.
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:76]: Installs and configures Oracle Container Runtime for Docker. We do not ship the Docker CE and Docker Engine is not actually a product name. I'd also make this a link to the documentation: https://docs.oracle.com/en/operating-systems/oracle-linux/docker/

With that being said, it'd would probably be better to change `docker-engine` to `container-runtime` to avoid potentialy trademark usage issues (Docker is a registered trademark).

File: OracleLinux/7/README.md [hunk 10]
```diff
@@ -1,36 +1,89 @@
...
+```
+
+Within the guest, run Docker commands, for example `docker run -it oraclelinux:7-slim` to run an Oracle Linux 7 container, or `docker run -ti oraclelinux:8-slim` to run an Oracle Linux 8 container
+
 ## Other info
 
-* If you need to, you can connect to the machine via `vagrant ssh`.
-* On the guest OS, the directory `/vagrant` is a shared folder and maps to wherever you have this file checked out.
+- If you need to, you can connect to the machine via `vagrant ssh`.
+- On the guest OS, the directory `/vagrant` is a shared folder and maps to wherever you have this file checked out.
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:89]: I'd rephrase this as "The directory in which the `Vagrantfile` is located is automatically mounted into the guest at `/vagrant` as a shared folder."

File: OracleLinux/7/scripts/docker-engine.sh [hunk 11]
```diff
@@ -0,0 +1 @@
+../../../DockerEngine/scripts/install.sh
```
Comment (Djelibeybi) [inline OracleLinux/7/scripts/docker-engine.sh:1]: I would copy the script into this location so that we can remove the `DockerEngine` and `LAMP` directories completely in a future pull request. I'd also rename this to `container-engine.sh`

File: OracleLinux/7/scripts/lamp.sh [hunk 12]
```diff
@@ -0,0 +1 @@
+../../../LAMP/scripts/lamp.sh
```
Comment (Djelibeybi) [inline OracleLinux/7/scripts/lamp.sh:1]: I would copy this script into the current location so we can remove the `LAMP` directory in the future.

File: OracleLinux/7/README.md [hunk 13]
```diff
@@ -1,36 +1,89 @@
...
+
+```shell
+EXTEND=lamp
+EXPOSE=8080:80
+```
+
+Guest port "80" (Apache) is redirected to Host port "8080".  
+Once ready, you can test it by opening following URL on your Host OS: http://localhost:8080/info.php
+
+### Docker engine
```
Comment (Djelibeybi) [inline OracleLinux/7/README.md:74]: Oracle Container Runtime for Docker

File: OracleLinux/7/.env [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+# in .env.local
+
+# Extensions
+# To extend the box, uncomment one of the following:
+
+# LAMP server
+# EXTEND=lamp
+# EXPOSE=8080:80
+
+# Docker engine
```
Comment (Djelibeybi) [inline OracleLinux/7/.env:15]: Oracle Container Runtime for Docker

File: OracleLinux/7/scripts/docker-engine.sh (hunk 11 above)
Comment (AmedeeBulle) [reply OracleLinux/7/scripts/docker-engine.sh:1]: I inverted the symlinks (to keep file history)

File: OracleLinux/7/README.md [hunk 15]
```diff
@@ -1,36 +1,89 @@
...
+
+- `EXTEND`: comma separated list of extensions.  
+   For each specified extension a script named `<extension>.sh` is automatically run during provisioning.  
+   Scripts must be located in the `scripts` or `scripts.local` directory.
+- `EXPOSE`: comma separated list of ports to expose in the format: `<host port>:<guest port>`
+
+Example: to extend the project with the `lamp` extension and expose the guest port 80 to 8080 on the host:
+
+```shell
+EXTEND=lamp EXPOSE=8080:80 vagrant up
```
Comment (PaulNeumann) [inline OracleLinux/7/README.md:53]: This won't work on Windows hosts. The Windows version would be:

```batch
set EXTEND=lamp && set EXPOSE=8080:80 && vagrant up
```

File: OracleLinux/7/README.md [hunk 16]
```diff
@@ -1,36 +1,89 @@
...
+
+### LAMP stack
+
+Provisions an Oracle Linux LAMP (**L**inux, **A**pache, **M**ySQL, **P**HP) stack.
+
+Usage:
+
+```shell
+EXTEND=lamp
+EXPOSE=8080:80
```
Comment (PaulNeumann) [inline OracleLinux/7/README.md:68]: Windows needs `set` before EXTEND and EXPOSE.

File: OracleLinux/7/README.md [hunk 17]
```diff
@@ -1,36 +1,89 @@
...
+Once ready, you can test it by opening following URL on your Host OS: http://localhost:8080/info.php
+
+### Oracle Container Runtime for Docker
+
+Installs and configures [Oracle Container Runtime for Docker](https://docs.oracle.com/en/operating-systems/oracle-linux/docker/) using Btrfs as storage.
+
+Usage:
+
+```shell
+EXTEND=container-runtime
```
Comment (PaulNeumann) [inline OracleLinux/7/README.md:81]: Windows needs `set` before EXTEND.

File: OracleLinux/7/README.md (hunk 15 above)
Comment (AmedeeBulle) [reply OracleLinux/7/README.md:53]: Good point...

I'll give an example for both for this one, and change the wording for the 2 others...

**Comment by Djelibeybi** [comment]
I'm approving this because it's good. I think we should revisit the use of SCL for the LAMP stack in another PR, now that we have newer versions of PHP available in dedicated repos that also have Oracle DB support (which the SCL version does not).
